pkg/limits/ingest_limits.go

```diff
@@ -376,35 +511,82 @@ func (s *IngestLimits) ServeHTTP(w http.ResponseWriter, _ *http.Request) {
 		Tenant          string   `json:"tenant"`
 		ActiveStreams   uint64   `json:"activeStreams"`
-		RecordedStreams []uint64 `json:"recordedStreams"`
+		AssignedStreams []uint64 `json:"assignedStreams"`
 	}
 
-	response := make(map[string]tenantLimits)
-	for tenant, streams := range s.metadata {
-		var activeStreams uint64
-		recordedStreams := make([]uint64, 0, len(streams))
+	// Get tenant and partitions from query parameters
+	params := r.URL.Query()
+	tenant := params.Get("tenant")
+	partitionsStr := params.Get("partitions")
+
+	var requestedPartitions []int32
+	if partitionsStr != "" {
+		// Split comma-separated partition list
+		partitionStrs := strings.Split(partitionsStr, ",")
+		requestedPartitions = make([]int32, 0, len(partitionStrs))
+
+		// Convert each partition string to int32
+		for _, p := range partitionStrs {
+			if val, err := strconv.ParseInt(strings.TrimSpace(p), 10, 32); err == nil {
+				requestedPartitions = append(requestedPartitions, int32(val))
+			}
+		}
+	}
 
-		// Count active streams and record their status
-		for hash, lastSeen := range streams {
-			isActive := lastSeen >= cutoff
-			if isActive {
-				activeStreams++
+	partitions := s.metadata[tenant]
+
+	var (
+		activeStreams   uint64
+		assignedStreams = make([]uint64, 0)
+		response        = make(map[string]tenantLimits)
+	)
+
+	for _, requestedID := range requestedPartitions {
+		// Consider the recorded stream if it's partition
+		// is one of the partitions we are still assigned to.
+		assigned := false
+		for assignedID := range partitions {
+			if requestedID == assignedID {
+				assigned = true
+				break
 			}
-			recordedStreams = append(recordedStreams, hash)
 		}
 
-		if activeStreams > 0 || len(recordedStreams) > 0 {
-			response[tenant] = tenantLimits{
-				Tenant:          tenant,
-				ActiveStreams:   activeStreams,
-				RecordedStreams: recordedStreams,
+		if !assigned {
+			continue
+		}
+
+		// If the stream is written into a partition we are
+		// assigned to and has been seen within the window,
+		// it is an active stream.
+		for _, stream := range partitions[requestedID] {
+			if stream.lastSeenAt >= cutoff {
+				activeStreams++
+				assignedStreams = append(assignedStreams, stream.hash)
 			}
 		}
 	}
 
+	if activeStreams > 0 || len(assignedStreams) > 0 {
+		response[tenant] = tenantLimits{
+			Tenant:          tenant,
+			ActiveStreams:   activeStreams,
+			AssignedStreams: assignedStreams,
+		}
+	}
+
 	util.WriteJSONResponse(w, response)
 }
 
+// GetAssignedPartitions implements the logproto.IngestLimitsServer interface.
+// It returns the partitions that the tenant is assigned to and the instance still owns.
+func (s *IngestLimits) GetAssignedPartitions(_ context.Context, _ *logproto.GetAssignedPartitionsRequest) (*logproto.GetAssignedPartitionsResponse, error) {
+	s.mtxAssingedPartitions.RLock()
+	defer s.mtxAssingedPartitions.RUnlock()
+
+	return &logproto.GetAssignedPartitionsResponse{AssignedPartitions: s.assingedPartitions}, nil
```

Comment:
Not sure if the lock is enough here you might need a copy since the marshaller will read after your lock.